src/storage/src/render/multi_worker_persist_sink.rs

```diff
@@ -0,0 +1,1123 @@
+                        new batch_description: {:?}",
+                    batch_description
+                );
+
+                output.give(&cap, batch_description).await;
+
+                // WIP: We downgrade our capability so that downstream
+                // operators (writer and appender) can know when all the
+                // writers have had a chance to write updates to persist for
+                // a given batch. Just stepping forward feels a bit icky,
+                // though.
+                let new_batch_frontier = Antichain::from_elem(batch_ts.step_forward());
+                trace!(
+                    "persist_sink {collection_id}/{shard_id}: \
+                        downgrading to {:?}",
+                    new_batch_frontier
+                );
+                let res = cap_set.try_downgrade(new_batch_frontier.iter());
+                match res {
+                    Ok(_) => (),
+                    Err(e) => panic!("in minter: {:?}", e),
+                }
+
+                // After successfully emitting a new description, we can update the upper for the
+                // operator.
+                current_upper = desired_frontier.to_owned();
+            }
+        }
+    });
+
+    if collection_id.is_user() {
+        output_stream.inspect(|d| trace!("batch_description: {:?}", d));
+    }
+
+    let token = Rc::new(shutdown_button.press_on_drop());
+    (output_stream, token)
+}
+
+/// Writes `desired_stream` to persist, but only for updates
+/// that fall into batch a description that we get via `batch_descriptions`.
+/// This forwards a `HollowBatch` (with additional metadata)
+/// for any batch of updates that was written.
+///
+/// This also and updates various metrics.
+fn write_batches<G>(
+    scope: &mut G,
+    collection_id: GlobalId,
+    operator_name: String,
+    target: &CollectionMetadata,
+    batch_descriptions: &Stream<G, (Antichain<Timestamp>, Antichain<Timestamp>)>,
+    desired_stream: &Stream<G, (Result<Row, DataflowError>, Timestamp, Diff)>,
+    persist_clients: Arc<PersistClientCache>,
+    storage_state: &mut StorageState,
+) -> (Stream<G, HollowBatchAndMetadata>, Rc<dyn Any>)
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let worker_index = scope.index();
+
+    let persist_location = target.persist_location.clone();
+    let shard_id = target.data_shard;
+    let target_relation_desc = target.relation_desc.clone();
+
+    let source_statistics = storage_state
+        .source_statistics
+        .get(&collection_id)
+        .expect("statistics initialized")
+        .clone();
+
+    let mut write_op =
+        AsyncOperatorBuilder::new(format!("{} write_batches", operator_name), scope.clone());
+
+    let (mut output, output_stream) = write_op.new_output();
+
+    let mut descriptions_input = write_op.new_input(&batch_descriptions.broadcast(), Pipeline);
+    let mut desired_input = write_op.new_input(desired_stream, Pipeline);
```

Comment:
If we assume that the data comes pre-sharded we should document that in the doc comments.